src/backend/distributed/serialize_distributed_ddls.c

```diff
@@ -0,0 +1,267 @@
+/*-------------------------------------------------------------------------
+ *
+ * serialize_distributed_ddls.c
+ *
+ * This file contains functions for serializing distributed DDLs.
+ *
+ * If you're adding support for serializing a new DDL, you should
+ * extend the following functions to support the new object class:
+ *   AcquireCitusAdvisoryObjectClassLockGetOid()
+ *   AcquireCitusAdvisoryObjectClassLockCheckPrivileges()
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "miscadmin.h"
+
+#include "catalog/dependency.h"
+#include "catalog/pg_database_d.h"
+#include "commands/dbcommands.h"
+#include "storage/lock.h"
+#include "utils/builtins.h"
+
+#include "pg_version_compat.h"
+
+#include "distributed/adaptive_executor.h"
+#include "distributed/argutils.h"
+#include "distributed/deparse_shard_query.h"
+#include "distributed/resource_lock.h"
+#include "distributed/serialize_distributed_ddls.h"
+
+
+PG_FUNCTION_INFO_V1(citus_internal_acquire_citus_advisory_object_class_lock);
+
+
+static void SerializeDistributedDDLsOnObjectClassInternal(ObjectClass objectClass,
+														  char *qualifiedObjectName);
+static char * AcquireCitusAdvisoryObjectClassLockCommand(ObjectClass objectClass,
+														 char *qualifiedObjectName);
+static void AcquireCitusAdvisoryObjectClassLock(ObjectClass objectClass,
+												char *qualifiedObjectName);
+static Oid AcquireCitusAdvisoryObjectClassLockGetOid(ObjectClass objectClass,
+													 char *qualifiedObjectName);
+static void AcquireCitusAdvisoryObjectClassLockCheckPrivileges(ObjectClass objectClass,
+															   Oid oid);
+
+
+/*
+ * citus_internal_acquire_citus_advisory_object_class_lock is an internal UDF
+ * to call AcquireCitusAdvisoryObjectClassLock().
+ */
+Datum
+citus_internal_acquire_citus_advisory_object_class_lock(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	PG_ENSURE_ARGNOTNULL(0, "object_class");
+	ObjectClass objectClass = PG_GETARG_INT32(0);
+
+	char *qualifiedObjectName = PG_ARGISNULL(1) ? NULL : PG_GETARG_CSTRING(1);
```

Comment:
Now this is based on "objectName" rather than "oid" because the previous approach was introducing a race condition where the subquery that we send to workers (to resolve the oid from the database) might return NULL oid in case another session concurrently dropped / renamed the object. And this was resulting in acquiring an incorrect lock (InvalidOid vs a real oid) even if it doesn't seem much harmful.


See https://github.com/citusdata/citus/actions/runs/7388502257/attempts/1#summary-20099571404.